packages/playground/website/src/components/ensure-playground-site/ensure-playground-site-slug.tsx

```diff
@@ -0,0 +1,78 @@
+import { useEffect } from 'react';
+import { useLocation } from 'wouter';
+import { urlContainsSiteConfiguration } from '../../lib/resolve-blueprint';
+import { listSites, SiteStorageTypes } from '../../lib/site-storage';
+import { useCurrentUrl, useSearchParams } from '../../lib/router-hooks';
+
+// @ts-ignore
+const opfsSupported = typeof navigator?.storage?.getDirectory !== 'undefined';
+
+// Treats the URL as the source of truth. Ensures we're always at a URL
+// that contains a site slug.
+export function EnsurePlaygroundSiteSlug({
+	children,
+}: {
+	children: React.ReactNode;
+}) {
+	const [query, setQuery] = useSearchParams();
+	const siteSlug = query.get('site-slug');
+	const [, setLocation] = useLocation();
+	const urlString = useCurrentUrl();
+
+	// Ensure the site slug is always present in the URL.
+	useEffect(() => {
+		/*
+		 * @TODO: Change the entire mental model of the `storage` parameter.
+		 *        For example, `storage=none` + an existing site slug makes
+		 *        no sense. We don't load where to load the site from, e.g.
+		 *        should it come from OPFS? Local directory? Network? We could
+		 * 		  separate the "load from" and "save" to operations, but they
+		 *        make more sense as user interactions than URL parameters.
+		 *        Perhaps we only need a single `load-site-from` URL parameter?
```

Comment:
Here are some thoughts...

It sounds like we're talking about the following three data flows:
- **Import**: Import site into local data store
- **Use**: Work with site in local data store
- **Export**: Export site from local data store

From Playground's perspective, all sites could be local, with manual or automated import/export operations as another layer.

As for the query API, it seems we have a couple of cases:
- **Create**: Create a site with a specific configuration
- **Load**: Identify a site to load and use its persisted configuration
    - Perhaps some query params can be allowed to tweak a site's config like using a different PHP or WP version, but for security and safety, maybe that should be a **create** operation that forks to a temporary site until the user explicitly saves over the original site.

What if the query API supports two explicit operations, create and load?

The **create** operation could do things like:
- Create a temporary site with or without blueprint
- Create a site that is stored in OPFS with or without blueprint
- Create a site that is stored on the local filesystem with or without blueprint
- Create a new local site based on a one-time import
- Create a new local site that updates itself based on an import each time it loads
- Create a new local site that knows how to save (i.e., export) to an external data store
 and is able to export to an external data store, either manually or automatically.

The **load** operation could be as simple as `https://playground.wordpress.net/?load=<uuid>`, and the site configuration could be looked up based on browser storage.

The only time the storage medium needs to matter is when a site is created. When we are loading sites, we can do that using a single identifier and look up the site configuration from whatever unified browser data store we choose.

Reply:
> It sounds like we're talking about the following three data flows:
> * Import: Import site into local data store
> * Use: Work with site in local data store
> * Export: Export site from local data store

I agree about the Import, Use, Export scenarios. At the same time, "local data store" seems too narrow. The three local stores we support today are:

* OPFS
* Local directories
* In-memory sites

Browsers keep evolving so we may have more in the future.

I would also like to support remote data sources like GitHub repos, remote zip files etc. These are remote, not local, and could fall under the importing data flow. I just want to acknowledge them. One use-case I have in mind is visiting, say, https://playground.wordpress.net/github/adamziel/blog, and getting a temporary site loaded from a GitHub repo.

Reply:
> As for the query API, it seems we have a couple of cases:
> * Create: Create a site with a specific configuration
> * Load: Identify a site to load and use its persisted configuration**
> What if the query API explicitly supports distinct create and load operations?

What would be the difference between the two? My mental model is Dockerfiles and base images. We provide minified WordPress as a default base image, but the API consumer is free to provide their own `wordpress.zip`. They may or may not provide Blueprint steps on top of that, and these steps may install plugins from a GitHub repo and even entirely replace the default WordPress files. I'm not sure how to distinguish between loading and creating a site.

Reply:
> > As for the query API, it seems we have a couple of cases:
> * Create: Create a site with a specific configuration
> * Load: Identify a site to load and use its persisted configuration**
> What if the query API explicitly supports distinct create and load operations?

I like it! One nuance is we'll still need to handle running an existing site with different runtime configuration, e.g. `let's turn off the networking and turn on more PHP extensions`. I'm now thinking we may remove the concept of specifying `?networking=yes` via URL for an existing site and always read that flag from the persistent store. This would mean we won't be able to change that configuration without interacting with the "site settings form", but maybe that's fine 🤔 It would be a bit less convenient for the developers, but it would save us merging a cascade of config values.

Reply:
Hm, I like your idea more and more @brandonpayton. 

I'm thinking about the state transition between `Create` and `Load` and it seems non-trivial. Let's consider a few scenarios:

* **Creating a new OPFS site.** Once we update the data store, we need to transition to a `Load` state. This requires updating the URL to ensure a page refresh won't create another site. What should we preserve in the URL? My gut says _only the site slug_. The rest of it should already be in the store, e.g. the Blueprint and the runtime configuration like networking and PHP extension bundles.
* **Creating a new MEMFS site.** When transitioning to a `Load` state we need to preserve _everything_ in the URL to make sure a page refresh creates an identical site. I'm not even sure if we want any site slug in the URL, as we'd try to resolve that site slug after a page refresh. The site wouldn't be there anymore, so we'd fail.
* **Saving a MEMFS site in OPFS.** This seems identical to "Creating a new OPFS site" scenario. We're just delaying the state transition until a specific user interaction.

Which means we'd get an inconsistent URL behavior based on the storage type. I'm not sure what to think about it yet. 

One alternative to get a more consistent behavior would be keeping all the query args and the blueprint in the URL for OPFS sites, and only prepending a `?site-slug=opfs:my-new-site` param. On the up side, it would preserve all the other parameters. On the down side, the user may easily miss the `?site-slug` part of the URL and get frustrated when their query arg adjustments don't work. We could detect that and explain how it works in the UI, although I'm worried how effective would that be – the current GitHub modal explains you're going to lose your data and even requires checking a checkbox, and it still confuses people.

We'd also have to decide what to do in cases like `https://playground.wordpress.net/site/opfs:my-site/?networking=yes`. We could either apply that on top of any stored configuration, or ignore these values. If we ignore them, an explicit warning in the UI would go a long way.

Tangentially, I like thinking about the query args as of CLI args – they're used to start the program and I can re-run the last command. Eventually we could share more names and code paths between the Playground website and Playground CLI.

Reply:
Actually, should we even support `?storage=browser`? It effectively says someone can create a persistent site in your store by just you clicking a link. That sounds invasive. Perhaps only supporting a temporary store via the Query API would be a better choice? And it would clear up a lot. I'm not sure how we could handle `?storage=local-fs` anyway.

Reply:
Removing the `storage` query arg would unlock the following flow:

1. Someone clicks on a Playground link
2. A temporary site is created and loaded, the URL stays unchanged
3. You can...
    * Refresh the page and get a new site created with the same recipe.
    * Switch to another site and get a different URL. We keep the `client` and `iframe` around so you continue seeing the temporary site in the site list and can return to it without losing your changes.
    * Click "store this site permanently". You're asked for a name or we pick a random one, and then the URL changes to point to the stored site. Blueprint or things like `?plugin=` are removed from the URL.

We also wouldn't have to figure out what `?storage=` means in terms of "where do I load it from" vs "where do I save it to". I'd say let's just do it. What do you think @brandonpayton?

Reply:
Oh! And this would mean we can assume all the sites with a slug are in `opfs`! And if there's one that actually lives in a local directory, the `metadata.json` / `playground.json` file could tell us to look for it there.

Reply:
I tried that in 1ad41dd16. The new feature flag is http://localhost:5400/website-server/?site-manager. I have about 30 minutes before I wrap up from this PR for the day, let's see where can I get it.

Reply:
> Actually, should we even support ?storage=browser? It effectively says someone can create a persistent site in your store by just you clicking a link. That sounds invasive.

Intuitively, this was a concern for me as well. It's a concern about a URL being able to decide too much for a user before they have a chance to consider it. It seems like we will come to a stronger sense of what is appropriate over time, and always starting with a temporary sites seems like a very safe place to start.

> Removing the `storage` query arg would unlock the following flow:
> 
> 1. You click on a Playground link.
> 2. You get a new temporary site. The URL stays unchanged. There's no way to create a persistent site through a link.
❤️   This sounds good and wise to me.

> 3. You can...
>    
>    * Refresh the page and get a new site created with the same recipe as the original site.
>    * Switch to another site and get a different URL. We keep the `client` and `iframe` around so you continue seeing the temporary site in the site list and can return to it without losing your changes.

Good stuff.

>    * Click "store this site permanently". You're asked for a name or we pick a random name for you, and then the URL changes to point to the stored site. Blueprints and things like `?plugin=` are removed from the URL.

One thing I wonder about is whether there is any security drawback to loading sites based on human-readable slugs that can be guessed (instead of just addressing persistent sites with randomly assigned UUIDs). Human-readable slugs are pleasant for humans IMO and possibly good for easy human URL recognition. Yet slugs might be more easily guessed by third parties who could craft links to them.

Whether that is a danger or not probably depends on whether a site-loading URL can trigger anything more than simply loading a site. If a URL can trigger loading and a site and running remote blueprint X, then third parties could manipulate sites by injecting things, deleting things, or whatever.

> In this case, opening Playground would still default to a temporary site like it does today. The only way to get a persistent site would be by explicitly clicking the "save" button. The subsequent visits to Playground would also default to a new temporary site. If that annoys the users, we might explore defaulting to the last used permanent site.

This makes sense to me and sounds generally good.

> [snip] 
> We also wouldn't have to figure out what `?storage=` means in terms of "where do I load it from" vs "where do I save it to". I'd say let's just do it. What do you think @brandonpayton?

Let's do it. :)

Reply:
> I tried that in [1ad41dd](https://github.com/WordPress/wordpress-playground/commit/1ad41dd165fd4b95aa636f70b0630c9e20955d3c). The new feature flag is http://localhost:5400/website-server/?site-manager. I have about 30 minutes before I wrap up from this PR for the day, let's see where can I get it.

Awesome. I'll play with it a bit now.

Reply:
> One nuance is we'll still need to handle running an existing site with different runtime configuration, e.g. let's turn off the networking and turn on more PHP extensions.

I wonder whether we should actually have first-class UI support for this use. It could be something like "Try with a different configuration" that prompts the user for config adjustments before launching a fork of the site.

> Tangentially, I like thinking about the query args as of CLI args – they're used to start the program and I can re-run the last command. Eventually we could share more names and code paths between the Playground website and Playground CLI.

👍 This seems like a good thought. If our concepts are well-defined, it seems like this could work out pretty well.

Reply:
So nice!

![Screenshot 2024-09-06 at 1 46 18 PM](https://github.com/user-attachments/assets/4526e7a1-a7bf-4b49-b8e8-9add1dcbd63a)

Reply:
I've spent a bit more time here and wow this simplified a lot. I've reduce the amount of types, code paths, and logic. I like this! I'm wrapping up this work for the weekend, feel free to take over @brandonpayton